packages/foundry/contracts/oracles/opyn/ZenBullOracle.sol

```diff
@@ -0,0 +1,176 @@
+    IZenBullStrategy public immutable zenBullStrategy;
+    IUniswapV3PoolState public immutable osqthWethPool;
+    IUniswapV3PoolState public immutable wethUsdcPool;
+    IERC20 public immutable eulerDToken;
+    IERC20 public immutable eulerEToken;
+    bytes6 public immutable usdcId;
+    bytes6 public immutable zenBullId;
+
+    event SourceSet(
+        ICrabStrategy crabStrategy,
+        IZenBullStrategy zenBullStrategy,
+        IUniswapV3PoolState osqthWethPool,
+        IUniswapV3PoolState wethUsdcPool,
+        IERC20 eulerDToken,
+        IERC20 eulerEToken
+    );
+
+    constructor(
+        ICrabStrategy crabStrategy_,
+        IZenBullStrategy zenBullStrategy_,
+        IUniswapV3PoolState osqthWethPool_,
+        IUniswapV3PoolState wethUsdcPool_,
+        IERC20 eulerDToken_,
+        IERC20 eulerEToken_,
+        bytes6 usdcId_,
+        bytes6 zenBullId_
+    ) {
+        crabStrategy = crabStrategy_;
+        zenBullStrategy = zenBullStrategy_;
+        osqthWethPool = osqthWethPool_;
+        wethUsdcPool = wethUsdcPool_;
+        eulerDToken = eulerDToken_;
+        eulerEToken = eulerEToken_;
+        usdcId = usdcId_;
+        zenBullId = zenBullId_;
+        emit SourceSet(
+            crabStrategy_,
+            zenBullStrategy_,
+            osqthWethPool_,
+            wethUsdcPool_,
+            eulerDToken_,
+            eulerEToken_
+        );
+    }
+
+    /**
+     * @notice Retrieve the value of the amount at the latest oracle price.
+     * Only `zenBullId` and `usdcId` are accepted as asset identifiers.
+     */
+    function peek(
+        bytes32 base,
+        bytes32 quote,
+        uint256 baseAmount
+    )
+        external
+        view
+        virtual
+        override
+        returns (uint256 quoteAmount, uint256 updateTime)
+    {
+        return _peek(base.b6(), quote.b6(), baseAmount);
+    }
+
+    /**
+     * @notice Retrieve the value of the amount at the latest oracle price. Same as `peek` for this oracle.
+     * Only `zenBullId` and `usdcId` are accepted as asset identifiers.
+     */
+    function get(
+        bytes32 base,
+        bytes32 quote,
+        uint256 baseAmount
+    )
+        external
+        virtual
+        override
+        returns (uint256 quoteAmount, uint256 updateTime)
+    {
+        return _peek(base.b6(), quote.b6(), baseAmount);
+    }
+
+    /**
+     * @notice Retrieve the value of the amount at the latest oracle price.
+     */
+    function _peek(
+        bytes6 base,
+        bytes6 quote,
+        uint256 baseAmount
+    ) private view returns (uint256 quoteAmount, uint256 updateTime) {
+        if (base == zenBullId && quote == usdcId) {
+            quoteAmount = (_getZenBullPrice() * baseAmount) / 1e18;
+        } else if (base == usdcId && quote == zenBullId) {
+            quoteAmount = (baseAmount * 1e18) / _getZenBullPrice();
+        } else {
+            revert ZenBullOracleUnsupportedAsset();
+        }
+        updateTime = block.timestamp;
```

Comment:
This is not true, is it? The updateTime would be the time of the last observation, if we can get that.